core/src/test/java/org/apache/iceberg/TestInternalData.java

```diff
@@ -0,0 +1,245 @@
+        InternalData.write(fileFormat, outputFile).schema(NESTED_SCHEMA).build()) {
+      appender.addAll(testData);
+    }
+
+    InputFile inputFile = fileIO.newInputFile(outputFile.location());
+    List<Record> readRecords = Lists.newArrayList();
+
+    try (CloseableIterable<Record> reader =
+        InternalData.read(fileFormat, inputFile)
+            .project(NESTED_SCHEMA)
+            .setCustomType(2, TestCustomRow.class)
+            .build()) {
+      for (Record record : reader) {
+        readRecords.add(record);
+      }
+    }
+
+    assertThat(readRecords).hasSize(testData.size());
+
+    for (int i = 0; i < testData.size(); i++) {
+      Record expected = testData.get(i);
+      Record actual = readRecords.get(i);
+
+      assertThat(actual.get(0, Long.class)).isEqualTo(expected.get(0, Long.class));
+
+      Object expectedNested = expected.get(1);
+      Object actualNested = actual.get(1);
+
+      if (expectedNested == null && actualNested == null) {
+        continue;
+      }
+
+      if (actualNested != null) {
+        assertThat(actualNested)
+            .as("Custom type should be TestCustomRow, but was: " + actualNested.getClass())
+            .isInstanceOf(TestCustomRow.class);
+        TestCustomRow customRow = (TestCustomRow) actualNested;
+        Record expectedRecord = (Record) expectedNested;
+
+        assertThat(customRow.get(0, Long.class))
+            .isEqualTo(expectedRecord.get(0, Long.class)); // inner_id
+        assertThat(customRow.get(1, String.class))
+            .isEqualTo(expectedRecord.get(1, String.class)); // inner_name
+      }
+    }
+  }
+
+  private List<Record> createSimpleTestRecords() {
+    Schema schema = simpleSchema();
+    List<Record> records = Lists.newArrayList();
+
+    Record record1 = GenericRecord.create(schema);
+    record1.set(0, 1L);
+    record1.set(1, "Alice");
+
+    Record record2 = GenericRecord.create(schema);
+    record2.set(0, 2L);
+    record2.set(1, "Bob");
+
+    records.add(record1);
+    records.add(record2);
+
+    return records;
+  }
+
+  private List<Record> createNestedTestRecords() {
+    List<Record> records = Lists.newArrayList();
+
+    Record record1 = GenericRecord.create(NESTED_SCHEMA);
+    record1.set(0, 1L);
+
+    Record nestedStruct1 =
+        GenericRecord.create(NESTED_SCHEMA.findType("nested_struct").asStructType());
+    nestedStruct1.set(0, 100L);
+    nestedStruct1.set(1, "inner_alice");
+    record1.set(1, nestedStruct1);
+
+    Record record2 = GenericRecord.create(NESTED_SCHEMA);
+    record2.set(0, 2L);
+    record2.set(1, null); // null nested struct
+
+    records.add(record1);
+    records.add(record2);
+
+    return records;
+  }
+
+  private Schema simpleSchema() {
+    return new Schema(
+        Types.NestedField.required(1, "id", Types.LongType.get()),
+        Types.NestedField.optional(2, "name", Types.StringType.get()));
+  }
+
+  public static class TestCustomRow implements StructLike {
```

Comment:
can we enhance the `CustomRow` from `TestHelpers` class by adding a new public constructor that takes a struct type?

Reply:
+1, This is possible.